xknx/io/tunnel.py

```diff
@@ -71,77 +74,117 @@ def init_udp_client(self):
         )
 
         self.udp_client.register_callback(
-            self.tunnel_reqest_received, [TunnellingRequest.service_type]
+            self._reqest_received,
+            [KNXIPServiceType.TUNNELLING_REQUEST, KNXIPServiceType.DISCONNECT_REQUEST],
         )
 
-    def tunnel_reqest_received(self, knxipframe, udp_client):
-        """Handle incoming tunnel request."""
-        # pylint: disable=unused-argument
-        if knxipframe.header.service_type_ident != KNXIPServiceType.TUNNELLING_REQUEST:
-            logger.warning("Service not implemented: %s", knxipframe)
-        else:
-            self.send_ack(
-                knxipframe.body.communication_channel_id,
-                knxipframe.body.sequence_counter,
-            )
-            # Don't handle invalid cemi frames (None) and only handle incoming L_DATA_IND frames.
-            # Ignore L_DATA_CON confirmation frames. L_DATA_REQ frames should only be outgoing.
-            if (
-                knxipframe.body.cemi is not None
-                and knxipframe.body.cemi.code is CEMIMessageCode.L_DATA_IND
-            ):
-                telegram = knxipframe.body.cemi.telegram
-                telegram.direction = TelegramDirection.INCOMING
-                if self.telegram_received_callback is not None:
-                    self.telegram_received_callback(telegram)
-
-    def send_ack(self, communication_channel_id, sequence_counter):
-        """Send tunnelling ACK after tunnelling request received."""
-        ack = TunnellingAck(
-            self.xknx,
-            communication_channel_id=communication_channel_id,
-            sequence_counter=sequence_counter,
-        )
-        self.udp_client.send(KNXIPFrame.init_from_body(ack))
+    ####################
+    #
+    # CONNECT DISCONNECT
+    #
+    ####################
 
     async def start(self):
```

Comment:
I like the idea. Since you now have dedicated start and stop functions it would probably make sense to add an async context manager to it and adjust the examples as well?

Reply:
I sure like the idea. But for now I aim for a stable and reliable connection. Since this is not very likely to be used directly I think syntactic sugar can be added later.